compatibility/checker.go

```diff
@@ -0,0 +1,404 @@
+	CheckDeploy(deploy *types.ServiceConfig) bool
+	CheckDeployEndpointMode(deploy *types.DeployConfig)
+	CheckDeployLabels(deploy *types.DeployConfig)
+	CheckDeployMode(deploy *types.DeployConfig)
+	CheckDeployReplicas(deploy *types.DeployConfig)
+	CheckDeployRestartPolicy(deploy *types.DeployConfig) bool
+	CheckDeployRollbackConfig(deploy *types.DeployConfig) bool
+	CheckDeployUpdateConfig(deploy *types.DeployConfig) bool
+	CheckPlacementConstraints(p *types.Placement)
+	CheckPlacementMaxReplicas(p *types.Placement)
+	CheckPlacementPreferences(p *types.Placement)
+	CheckRestartPolicyDelay(policy *types.RestartPolicy)
+	CheckRestartPolicyCondition(policy *types.RestartPolicy)
+	CheckRestartPolicyMaxAttempts(policy *types.RestartPolicy)
+	CheckRestartPolicyWindow(policy *types.RestartPolicy)
+	CheckUpdateConfigDelay(rollback string, config *types.UpdateConfig)
+	CheckUpdateConfigFailureAction(rollback string, config *types.UpdateConfig)
+	CheckUpdateConfigMaxFailureRatio(rollback string, config *types.UpdateConfig)
+	CheckUpdateConfigMonitor(rollback string, config *types.UpdateConfig)
+	CheckUpdateConfigOrder(rollback string, config *types.UpdateConfig)
+	CheckUpdateConfigParallelism(rollback string, config *types.UpdateConfig)
+	CheckDeployResourcesNanoCPUs(s string, resource *types.Resource)
+	CheckDeployResourcesMemoryBytes(s string, resource *types.Resource)
+	CheckDeployResourcesGenericResources(s string, resource *types.Resource)
+	CheckDeployResourcesLimits(deploy *types.DeployConfig) bool
+	CheckDeployResourcesReservations(deploy *types.DeployConfig) bool
+	CheckHealthCheck(service *types.ServiceConfig) bool
+	CheckLogging(service *types.ServiceConfig) bool
+	CheckNetworks(service *types.ServiceConfig) bool
+	CheckPorts(service *types.ServiceConfig) bool
+	CheckServiceVolumes(service *types.ServiceConfig) bool
+	CheckNetworkConfigIpam(network *types.NetworkConfig)
+	CheckNetworkConfigDriver(network *types.NetworkConfig)
+	CheckNetworkConfigDriverOpts(network *types.NetworkConfig)
+	CheckNetworkConfigExternal(network *types.NetworkConfig)
+	CheckNetworkConfigInternal(network *types.NetworkConfig)
+	CheckNetworkConfigAttachable(network *types.NetworkConfig)
+	CheckNetworkConfigLabels(network *types.NetworkConfig)
+}
+
+func Check(project *types.Project, c Checker) {
+	for i, service := range project.Services {
+		CheckServiceConfig(&service, c)
+		project.Services[i] = service
+	}
+
+	for i, network := range project.Networks {
+		CheckNetworkConfig(&network, c)
+		project.Networks[i] = network
+	}
+
+	for i, volume := range project.Volumes {
+		CheckVolumeConfig(&volume, c)
+		project.Volumes[i] = volume
+	}
+
+	for i, config := range project.Configs {
+		CheckConfigsConfig(&config, c)
+		project.Configs[i] = config
+	}
+
+	for i, secret := range project.Secrets {
+		CheckSecretsConfig(&secret, c)
+		project.Secrets[i] = secret
+	}
+}
+
+func CheckServiceConfig(service *types.ServiceConfig, c Checker) {
```

Comment:
why arent these methods of Checker?

Reply:
Checker defines the per-attribute validation rule.
Those methods iterate on all attributes to call the adequate Checker

Reply:
With an Object-Oriented approach I would actually have set `CheckService` as part of the Default implementation, which would invoke sub Checkers, and give an option for implementors to override and skip/re-implement the whole service checking on it's own. But this is Golang